cli/tests/watcher_tests.rs

```diff
@@ -0,0 +1,1247 @@
+    write(&another_test, "syntax error ^^").unwrap();
+    assert_contains!(stderr_lines.next().unwrap(), "Restarting");
+    assert_contains!(stderr_lines.next().unwrap(), "error:");
+    assert_contains!(stderr_lines.next().unwrap(), "Test failed");
+
+    // Then restore the file
+    write(&another_test, "Deno.test('another one', () => 3 + 3)").unwrap();
+    assert_contains!(stderr_lines.next().unwrap(), "Restarting");
+    assert_contains!(stdout_lines.next().unwrap(), "running 1 test");
+    assert_contains!(stdout_lines.next().unwrap(), "another one");
+    stdout_lines.next();
+    stdout_lines.next();
+    stdout_lines.next();
+    wait_contains("Test finished", &mut stderr_lines);
+
+    // Confirm that the watcher keeps on working even if the file is updated and the test fails
+    // This also confirms that it restarts when dependencies change
+    write(
+      &foo_file,
+      "export default function foo() { throw new Error('Whoops!'); }",
+    )
+    .unwrap();
+    assert_contains!(stderr_lines.next().unwrap(), "Restarting");
+    assert_contains!(stdout_lines.next().unwrap(), "running 1 test");
+    assert_contains!(stdout_lines.next().unwrap(), "FAILED");
+    wait_for(|m| m.contains("FAILED"), &mut stdout_lines);
+    stdout_lines.next();
+    wait_contains("Test finished", &mut stderr_lines);
+
+    // Then restore the file
+    write(&foo_file, "export default function foo() { 1 + 1 }").unwrap();
+    assert_contains!(stderr_lines.next().unwrap(), "Restarting");
+    assert_contains!(stdout_lines.next().unwrap(), "running 1 test");
+    assert_contains!(stdout_lines.next().unwrap(), "foo");
+    stdout_lines.next();
+    stdout_lines.next();
+    stdout_lines.next();
+    wait_contains("Test finished", &mut stderr_lines);
+
+    // Test that circular dependencies work fine
+    write(
+      &foo_file,
+      "import './bar.js'; export default function foo() { 1 + 1 }",
+    )
+    .unwrap();
+    write(
+      &bar_file,
+      "import './foo.js'; export default function bar() { 2 + 2 }",
+    )
+    .unwrap();
+    check_alive_then_kill(child);
+  }
+
+  #[flaky_test]
+  fn test_watch_doc() {
+    let t = TempDir::new();
+
+    let mut child = util::deno_cmd()
+      .current_dir(util::testdata_path())
+      .arg("test")
+      .arg("--watch")
+      .arg("--doc")
+      .arg("--unstable")
+      .arg(t.path())
+      .env("NO_COLOR", "1")
+      .stdout(std::process::Stdio::piped())
+      .stderr(std::process::Stdio::piped())
+      .spawn()
+      .unwrap();
+    let (mut stdout_lines, mut stderr_lines) = child_lines(&mut child);
+
+    assert_eq!(stdout_lines.next().unwrap(), "");
+    assert_contains!(stdout_lines.next().unwrap(), "0 passed | 0 failed");
+    wait_contains("Test finished", &mut stderr_lines);
+
+    let foo_file = t.path().join("foo.ts");
+    write(
+      &foo_file,
+      r#"
+    export default function foo() {}
+  "#,
```

Comment:
This indentation has now added extra indent to these string. Not a big deal though.